infrastructure/comms/src/connection/peer_connection/runner.rs

```diff
@@ -0,0 +1,125 @@
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::{
+    sync::mpsc::{sync_channel, RecvTimeoutError, SyncSender},
+    thread,
+    time::Duration,
+};
+
+use crate::connection::{
+    connection::Connection,
+    message::Frame,
+    peer_connection::{control::ControlMessage, PeerConnectionContext, PeerConnectionError},
+    types::Direction,
+    ConnectionError,
+};
+
+macro_rules! try_recv {
+    ($inner: expr) => {
+        match $inner {
+            Ok(f) => Some(f),
+            Err(e) => match e {
+                ConnectionError::Timeout => None,
+                _ => break Err(e),
+            },
+        }
+    };
+}
+
+/// Take the given PeerConnectionContext and use it to start a worker thread which:
+/// - Establishes a connection to peer
+/// - Establishes a connection to the message consumer
+/// - Receives and handles ControlMessages
+///
+/// # Arguments
+///
+/// `context` - The PeerConnectionContext which will be owned by the worker thread
+pub(super) fn start_thread(context: PeerConnectionContext) -> SyncSender<ControlMessage> {
+    let (sender, receiver) = sync_channel(5);
+
+    let mut identity: Option<Frame> = None;
+
+    thread::spawn(move || {
+        let peer_conn = Connection::new(&context.context, context.direction.clone())
+            .set_curve_encryption(context.curve_encryption)
+            .set_receive_hwm(10)
+            .set_send_hwm(10)
+            .set_socks_proxy_addr(context.socks_address)
+            .set_max_message_size(Some(context.max_msg_size))
+            .establish(&context.peer_address)?;
+
+        let consumer = Connection::new(&context.context, Direction::Outbound).establish(&context.consumer_address)?;
```

Comment:
This is in the thread, is the behaviour of the "?" not different?

Reply:
This is one of the issues, the thread can return a result when you join to it (`JoinHandle<Result<(), ConnectionError>>`) but we never use it. I still have to think about what we should do if the thread returns an error - but likely we'll "tell" some other component through zmq Pub/Sub and update our connection state. Further iterations will address this